src/webui/www/tstool.py

```diff
@@ -0,0 +1,192 @@
+    
+    root = tree.getroot()
+    sources = copy.deepcopy(source_ts)
+    for context in root.findall('context'):
+        context_name = context.find('name').text
+        has_context = context_name in sources
+        if not has_context and no_obsolete:
+            root.remove(context)
+            continue
+        
+        for message in context.findall('message'):
+            for location in message.findall('location'):
+                message.remove(location)
+                
+            source = message.find('source').text
+            translation = message.find('translation')
+            if has_context and source in sources[context_name]:
+                sources[context_name].remove(source)
+                    
+                trtype = translation.attrib.get('type')
+                if (trtype == 'obsolete') or (trtype == 'vanished'):
+                    del translation.attrib['type'] # i.e. finished
+            else:
+                if no_obsolete or (translation.attrib.get('type', '') == 'unfinished'):
+                    context.remove(message)
+                else:
+                    translation.attrib['type'] = 'vanished'
+
+        if not has_context:
+            continue
+
+        # add new messages for current context
+        for source in sources[context_name]:
+            message = ET.SubElement(context, 'message')
+            ET.SubElement(message, 'source').text = source
+            ET.SubElement(message, 'translation', {'type': 'unfinished'})
+        del sources[context_name]
+
+    # add messages for new contexts
+    for context_name in sources:
+        context = ET.SubElement(root, 'context')
+        ET.SubElement(context, 'name').text = context_name
+
+        for source in sources[context_name]:
+            message = ET.SubElement(context, 'message')
+            ET.SubElement(message, 'source').text = source
+            ET.SubElement(message, 'translation', {'type': 'unfinished'}) 
+
+    # prettify output xml
+    indent = ' ' * 4
+    root.text = '\n'
+    for context in root.findall('./context'):
+        context.text = '\n' + indent
+        context.tail = '\n'
+        context.find('./name').tail = '\n' + indent
+        messages = context.findall('./message')
+        if len(messages) == 0: continue
+        
+        for message in messages:
+            message.text = '\n' + (indent * 2)
+            message.tail = '\n' + indent
+            elems = message.findall('./')
+            if len(elems) == 0: continue
+            
+            for elem in elems:
+                elem.tail = '\n' + (indent * 2)
+            elems[-1:][0].tail = '\n' + indent
+        messages[-1:][0].tail = '\n'
+
+    try:
+        file = open(filename, mode = 'wb')
+        file.write(b'<?xml version="1.0" encoding="utf-8"?>\n'
+                   b'<!DOCTYPE TS>\n')
+        tree.write(file, encoding = 'utf-8')
+    except:
+        print('\tFailed to write %s!' % (normpath(filename)))
```

Comment:
If the exception is unexpected... I think the pythonic way is not catching it, just let the program terminates. Otherwise you should catch the specific exception.
ps. this also applies to the try...catch above.

Reply:
>If the exception is unexpected... I think the pythonic way is not catching it, just let the program terminates.

Why should I allow the program to terminate in case it is impossible to write some specific file for some reason? And Yes, I'm not at all interested in that reason. I don't want to make a monster out of this simple helper script. It is not intended for general use, but only for developers.

P.S. If someone does not have enough problems in the main qBittorrent code base, or someone will find here some critical (in his opinion) shortcomings, he can spend his time and bring it to perfection later.

Reply:
>Problem is you're catching all exceptions, you should only catch the ones you need

In general case, Yes. But as I said before:

> And Yes, I'm not at all interested in that reason.

I don't know what exceptions there can be and I don't want to know it exactly. I just want to skip it and try to continue.
It works without errors under normal conditions. I don't want to complicate it trying to handle abnormal environment state. I assume the developer who encounter some error will be able to handle it.

Reply:
>Then maybe try block is not really needed and can be omitted?

You didn't still understand desired behavior? Skip any error (only print message about it) and try to process the rest files. I can't implement it without try-catch.
Of course, I can omit try-catch in this place, since error here usually means some external problem breaking the correct execution of the script. But I can't omit it above, since there can be some particular invalid file which can be skipped.